### PR TITLE
Add back docker caching, simplify docker builds

### PR DESCRIPTION
**Pull request by rbren**
**What is the problem that this fixes or functionality that this introduces? Does it fix any open issues?**

[This PR](https://github.com/All-Hands-AI/OpenHands/pull/2456) made a bunch of changes to the image workflow. Some of the changes are great, but we lost a few things as well:
* we stopped pushing image tags for PRs (which is helpful for QA)
* we lost caching
* multi-arch became part of a matrix, leading (I think) to longer build times

---
**Give a summary of what the PR does, explaining any non-trivial design decisions**

* Changes `build.sh` to build a single multi-arch image, rather than iterating through each platform
* Adds back the `--push` option to the build script
* Removes the `.tar` image files and artifacts, as well as a lot of plumbing for those images and tag names

---
**Other references**

### CODE REVIEW

**Comment by rbren** [comment]
Whoops thanks Tobi--I added all the tests back. They'll have slightly different names though due to the matrix changes

File: containers/build.sh [hunk 1]
```diff
@@ -3,13 +3,24 @@ set -eo pipefail
...
 org_name=$2
-platform=$3
+push=0
+if [[ $3 == "--push" ]]; then
+  push=1
+fi
+tag_suffix=$4
 
-echo "Building: $image_name for platform: $platform"
-tags=()
```
Comment (xingyaoww) [inline containers/build.sh:9]: Is this `tags=()` intentionally removed? Or it is a mistake?

File: .github/workflows/ghcr_runtime.yml [hunk 2]
```diff
@@ -72,77 +74,17 @@ jobs:
...
-          pattern: last-tag-*
-          path: /tmp/
-          merge-multiple: true
-      - name: Set up test matrix
-        id: set-matrix
-        run: |
-          matrix=$(cat /tmp/last-tag-*.txt | sort -u | jq -R -s -c 'split("\n") | map(select(length > 0))')
-          echo "test_image_tags=$matrix" >> $GITHUB_OUTPUT
-          echo "Generated test_image_tags: $matrix"
+          suffix=$(echo "${{ matrix.base_image }}" | cut -d ':' -f 1 | cut -d '/' -f 1)
```
Comment (xingyaoww) [inline .github/workflows/ghcr_runtime.yml:77]: We might need better approach than this going forward, but probably good for now

File: containers/build.sh [hunk 3]
```diff
@@ -3,13 +3,24 @@ set -eo pipefail
...
+echo "Building: $image_name"
```
Comment (xingyaoww) [inline containers/build.sh:12]:
```suggestion
echo "Building: $image_name"
tags=()
```

**Review by xingyaoww — APPROVED** [review]
Overall LGTM! Couple nits

File: .github/workflows/ghcr_runtime.yml (hunk 2 above)
Comment (rbren) [reply .github/workflows/ghcr_runtime.yml:77]: yeah agreed

File: containers/build.sh (hunk 1 above)
Comment (rbren) [reply containers/build.sh:9]: good catch!

**Pull request merged**